### ISSUE TITLE
DbUpdateConcurrencyException when using sequences

### ISSUE THREAD

**Issue by mcgiany** [aspnet/Identity#1103]
Hi,

i'm using Identity with sequences instead of auto increment ID. The problem is, that i randomly gets DbUpdateConcurrencyException when i try to create User.

This is my sequence:
```
CREATE SEQUENCE [backoffice].[seqAspNetUsers]
		AS BIGINT
		START WITH 100
		INCREMENT BY 1
		NO MAXVALUE
		NO CYCLE
		CACHE 10
```

This is how i bind sequences to Indetity tables:

```
protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);
            // override schema, create all table in schema backoffice
            builder.HasDefaultSchema("backoffice");
            builder.HasSequence<int>("seqAspNetUsers", "backoffice").IncrementsBy(1);
            builder.HasSequence<int>("seqAspNetRoles", "backoffice").IncrementsBy(1);
            builder.HasSequence<int>("seqAspNetRoleClaims", "backoffice").IncrementsBy(1);
            builder.HasSequence<int>("seqAspNetUserClaims", "backoffice").IncrementsBy(1);

            builder.Entity<BackofficeUser>(b =>
                {
                    b.Property(e => e.Id).ForSqlServerUseSequenceHiLo("seqAspNetUsers", "backoffice");
                });

            builder.Entity<BackofficeRole>(b =>
                {
                    b.Property(e => e.Id).ForSqlServerUseSequenceHiLo("seqAspNetRoles", "backoffice");
                });

            builder.Entity<IdentityRoleClaim<int>>(b =>
                {
                    b.Property(e => e.Id).ForSqlServerUseSequenceHiLo("seqAspNetRoleClaims", "backoffice");
                });

            builder.Entity<IdentityUserClaim<int>>(b =>
                {
                    b.Property(e => e.Id).ForSqlServerUseSequenceHiLo("seqAspNetUserClaims", "backoffice");
                });
        }
```

Stack trace:
```
Microsoft.EntityFrameworkCore.DbUpdateConcurrencyException: Database operation expected to affect 1 row(s) but actually affected 0 row(s). Data may have been modified or deleted since entities were loaded. See http://go.microsoft.com/fwlink/?LinkId=527962 for information on understanding and handling optimistic concurrency exceptions.
at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ThrowAggregateUpdateConcurrencyException(Int32 commandIndex, Int32 expectedRowsAffected, Int32 rowsAffected)
at async Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ConsumeResultSetWithoutPropagationAsync(?)
at async Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ConsumeAsync(?)
at async Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(?)
at async Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(?)
at async Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(?)
at async Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(?)
at async Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(?)
at async Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore`8.CreateAsync[](?)
at async Microsoft.AspNetCore.Identity.UserManager`1.CreateAsync[](?)
at async Microsoft.AspNetCore.Identity.UserManager`1.CreateAsync[](?)
at async Neo.Backoffice.Business.Implementation.UserManagerStore.CreateAsync(?) in C:\agent2\_work\7\s\Neo.Backoffice\Business\Implementation\UserManagerStore.cs:line 44
at async Neo.Backoffice.Controllers.UsersController.CreatePost(?) in C:\agent2\_work\7\s\Neo.Backoffice\Controllers\UsersController.cs:line 192
at async Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionFilterAsync(?)
at async Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeAsync(?)
at async Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(?)
at async Neo.Backoffice.Middlewares.CheckUserStatusMiddleware.Invoke(?) in C:\agent2\_work\7\s\Neo.Backoffice\Middlewares\CheckUserStatusMiddleware.cs:line 40
at async Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(?)
at async Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.Invoke[](?)
at async Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(?)
at async Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(?)
```

I can see in the database, that the sequence is increased by 1, everytime i call UserStore.CreateAsync, but still getting the exception. Highest ID in AspNetUsers table is 8, so there is no problem with duplicate ID. When i recycle the pool, the exception disappear and creating works fine. And from time to time it comes back again, and i dont know why.

thanks

**Comment by divega** [comment]
@mcgiany could you share the code of `UsersController.CreatePost()`?

**Comment by divega** [comment]
@mcgiany Assuming `CreatePost()` results in a simple call to `UserStore.CreateAsync()` this could be some kind of race condition, but it is hard to say where.

My first suspicion is a similar failure we observed in our tests long time ago while running against SQL Server LocalDB. We were then able to identify root cause as an issue in LocalDB itself (i.e. it reproduced without EF Core in the picture). We believe the issue has since then been fixed but you might need a build of SQL Server that contains the fix. 

Talking about this issue with @ajcvickers he also suggested there could be something in our Hi/Lo key generation feature. 

For full context, `ForSqlServerUseSequenceHiLo()` won't really bind the default value of the column to the next value of the sequence directly, but will setup a key generation strategy that performs small increments in memory and periodically performs larger increments in the database (see http://stackoverflow.com/questions/282099/whats-the-hi-lo-algorithm for more complete explanations of how Hi/Lo works). 

There is really no point in using Hi/Lo with an increment of 1 on the database, so presumably we haven't tested this corner case enough and it is theoretically possible that it could cause a race condition.

In summary, here are a few questions and suggestions on what to try:

1. What database product are you using? Does it happen to be SQL Server LocalDB? Are you running on the latest version?

2. Could you try switching to increments of 2 or more and see if the issue goes away?

3. Can you try switching to binding the default value of the column to the next value of the sequence using this pattern to see if that helps?:

``` C#
    b.Property(e => e.Id).HasDefaultValueSql("NEXT VALUE FOR seqAspNetUsers");
```

**Comment by mcgiany** [comment]
Thanks @divega for your reaction.
This is code of CreatePost method:

```
[HttpPost]
        [ActionName("Create")]
        public async Task<IActionResult> CreatePost()
        {
            var model = await BindViewModel<UsersCreateViewModel>();
            model.UpdateRoles(_roleManager.Roles.ToList());

            if (!ModelState.IsValid)
            {
                return View(model);
            }

            BackofficeUser user = new BackofficeUser()
            {
                Name = model.Name,
                IsEnabled = true,
                Description = model.Description,
                Email = model.Email,
                UserName = model.Username
            };

            using (var transaction = _context.Database.BeginTransaction())
            {
                IdentityResult result = await _userManagerStore.CreateAsync(user, model.Password);
                if (result.Succeeded && model.SelectedRoles != null && model.SelectedRoles.Count > 0)
                {
                    result = await _userManagerStore.AddToRolesAsync(user, model.SelectedRoles);
                }

                if (result.Succeeded)
                {
                    transaction.Commit();
                    AddStatusSuccess(Strings.Users.UserCreatedMessage(user.UserName));

                    return RedirectToAction("Index");
                }

                transaction.Rollback();

                foreach (var error in result.Errors)
                {
                    ModelState.AddModelError(string.Empty, error.Description);
                }

                return View(model);
            }
        }
```
Answer to other questions:
1. I'm using latest SQL Server 2016
2. I could. I will tell you if it helps, or if the issue still exists. It can take some time, cuz the issue happened randomly.
3. This is not an option for me, because the table will be used in replication. Thats why i use hi/lo sequence with 1 increment.

**Comment by mcgiany** [comment]
OK, i changed the model and use sequences as default value. For this time, i think its enough. If i have some problem with this setup, i will reopen the issue.
Thanks again
